tesseract-server/src/logic_layer/config.rs

```diff
@@ -138,4 +158,82 @@ impl LogicLayerConfig {
             None => cut
         }
     }
+
+    /// Returns a unique name definition for a given level if there is one.
+    pub fn find_unique_level_name(&self, level_name: String) -> Option<&str> {
+        if let Some(name_substitutions) = &self.name_substitutions {
+            if let Some(levels) = &name_substitutions.levels {
+                for level in levels {
+                    if level.level == level_name {
+                        return Some(&level.unique_name)
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    /// Returns a unique name definition for a given property if there is one.
+    pub fn find_unique_property_name(&self, property_name: String) -> Option<&str> {
+        if let Some(name_substitutions) = &self.name_substitutions {
+            if let Some(properties) = &name_substitutions.properties {
+                for property in properties {
+                    if property.property == property_name {
+                        return Some(&property.unique_name)
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    /// Ensures level and property names are unique inside each cube based on
+    /// name substitutions from a logic layer configuration.
+    pub fn has_unique_levels_properties(&self, schema: &Schema) -> bool {
+        for cube in &schema.cubes {
+            let mut levels = HashSet::new();
+            let mut properties = HashSet::new();
+
+            for dimension in &cube.dimensions {
+                for hierarchy in &dimension.hierarchies {
+
+                    // Check each cube for unique level and property names
+                    for level in &hierarchy.levels {
+                        let level_name = format!(
```

Comment:
Looks like you're matching on the string of the fully qualified name (I think? let me know if this isn't true).

It would be more ergonomic for the user if the `level` field they filled in in the config got parsed to a `LevelName` and then you could check that `LevelName` are equal.

The benefits would be giving an error on attempting to parse an ill-formed name, and allowing the user to enter a non-qualified name (just dim and level).

Reply:
Oh, this would also allow not using brackets in the name.